x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java

```diff
@@ -67,6 +77,47 @@ public FieldAttribute(
         this.field = field;
     }
 
+    @SuppressWarnings("unchecked")
+    public <S extends StreamInput & PlanStreamInput> FieldAttribute(StreamInput in) throws IOException {
+        /*
+         * The funny casting dance with `<S extends...>` and `(S) in` is required
+         * because we're in esql-core here and the real PlanStreamInput is in
+         * esql-proper. And because NamedWriteableRegistry.Entry needs StreamInput,
+         * not a PlanStreamInput. And we need PlanStreamInput to handle Source
+         * and NameId. This should become a hard cast when we move everything out
+         * of esql-core.
+         */
+        this(
+            Source.readFrom((S) in),
```

Comment:
Oh no, having to do the casting dance is a bit frustrating :/

But! At least we can lose the generic parameter; we should also document that `in` must be a `PlanStreamInput`.

nit:
```suggestion
    /**
     * @param in must also implement {@link PlanStreamInput}
     */
    @SuppressWarnings("unchecked")
    public FieldAttribute(StreamInput in) throws IOException {
        /*
         * The funny casting dance with `(StreamInput & PlanStreamInput) in` is required
         * because we're in esql-core here and the real PlanStreamInput is in
         * esql-proper. And because NamedWriteableRegistry.Entry needs StreamInput,
         * not a PlanStreamInput. And we need PlanStreamInput to handle Source
         * and NameId. This should become a hard cast when we move everything out
         * of esql-core.
         */
        this(
            Source.readFrom((StreamInput & PlanStreamInput) in),
```

Reply:
Oh! That's cleaner. I honestly had no idea you could double cast like that.